File: src/main/java/com/leet/medium/SortColors.java
package com.leet.medium;

class Solution {
  void swap(int[] nums, int a, int b) {
    int tmp = nums[a];
    nums[a] = nums[b];
    nums[b] = tmp;
  }

  public void sortColors(int[] nums) {
    int zero = 0, two = nums.length - 1, ptr = 0;
    while (ptr <= two) {
      switch (nums[ptr]) {
      case 0:
        swap(nums, zero++, ptr++);
        break;
      case 1:
        ptr++;
        break;
      case 2:
        swap(nums, two--, ptr);
        break;
      }
    }
  }
}

File: src/main/java/com/leet/medium/SortColors.ts
/**
 Do not return anything, modify nums in-place instead.
 */
const RED = 0;
const WHITE = 1;
const BLUE = 2;

function sortColors(nums: number[]): void {
  let red = 0, blue = nums.length - 1, ptr = 0;

  const swap = (x: number, y: number) => {
    const tmp = nums[x];
    nums[x] = nums[y];
    nums[y] = tmp;
  };

  while (ptr <= blue) {
    if (nums[ptr] === BLUE) {
      swap(ptr, blue--);
    } else if (nums[ptr] === WHITE) {
      ptr++;
    } else if (nums[ptr] == RED) {
      swap(ptr++, red++);
    }
  }
}
